Added wallet connect button

pauljickling opened the pull request:
### What was wrong?

Copying and pasting wallet addresses is a pain. So a button is added that connects to web3 to get wallet address, and inserts it into the input field.

### How was it fixed?

Hopefully it is fixed! However I had no good way to test the wallet connection while running Pretix locally, so the moment of truth will be on the staging server.

#### Cute Animal Picture

![https://i.pinimg.com/564x/9c/b0/b5/9cb0b50043c4f9f952a3fa6d400c1d2c.jpg](https://i.pinimg.com/564x/9c/b0/b5/9cb0b50043c4f9f952a3fa6d400c1d2c.jpg)

pauljickling committed:
Added wallet connect button

skylarweaver commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -1,7 +1,143 @@
+  // Purge UI elements any previously loaded accounts
+  accountContainer.innerHTML = '';
+
+  // Go through all accounts and get their ETH balance
+  const rowResolvers = accounts.map(async (address) => {
+    const clone = template.content.cloneNode(true);
+    clone.querySelector(".address").textContent = address;
+    accountContainer.appendChild(clone);
+  });
```
We do not need their ETH balance right?

ligi replied:
yes - we do not need the balance - it is just about the address

skylarweaver commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -31,4 +167,13 @@ $(function () {
+  const formGroupElements = document.querySelectorAll("div.form-group");
+
+  const lastFormGroupElement = formGroupElements.item(formGroupElements.length - 1);
+  const inputElement = document.querySelector("input[name='3-question_1']");
```
Where is this used?

skylarweaver commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -1,7 +1,143 @@
+  const web3 = new Web3(provider);
+
+  // Get connected chain id from Ethereum node
+  const chainId = await web3.eth.getChainId();
+  // Load chain information over an HTTP API
+  const chainData = await EvmChains.getChain(chainId);
+  document.querySelector("#network-name").textContent = chainData.name;
```
Do we want chain ID (Mainnet, etc)? Do we display that anywhere? Prob not...

ligi replied:
no  we do not want the chainID

skylarweaver commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -1,7 +1,143 @@
+    return;
+  }
+  
+  const web3 = new Web3(provider);
+  const accounts = await web3.eth.getAccounts();
+  document.querySelector('#id_3-question_1').value = accounts[0];
```
I feel like this should be done in the refreshAccounrData() function. Maybe I'm wrong though

ligi replied:
yes

pauljickling replied:
That makes sense to me.

kvbik replied:
also instead of `id_3-question_1` you need to use `get_payment_eth_question_ids()` to filter the question field IDs (in the simplest case it will be just one) and fill the address there.

skylarweaver commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -1,7 +1,143 @@
+  // MetaMask does not give you all accounts, only the selected account
+  selectedAccount = accounts[0];
+
+  document.querySelector("#selected-account").textContent = selectedAccount;
```
Is this where we should be setting the address input value?

pauljickling replied:
That looks like more leftover copy/paste cruft. Will remove it.

skylarweaver commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -1,7 +1,143 @@
+
+  document.querySelector("#selected-account").textContent = selectedAccount;
+
+  // Get a handl
+  const template = document.querySelector("#template-balance");
+  const accountContainer = document.querySelector("#accounts");
+
+  // Purge UI elements any previously loaded accounts
+  accountContainer.innerHTML = '';
```
Do we have any of these html elements on our page? Not sure if we should or if this is just copy/paste

skylarweaver commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -1,7 +1,143 @@
+      package: WalletConnectProvider,
+    },
+
+    fortmatic: {
+      package: Fortmatic,
+    }
```
Do we want to use Formatic as a provider? Question for @ligi probably. If so, it looks like we need a key (prob a developer key we obtain from their site?):
![image](https://user-images.githubusercontent.com/4008777/80798933-6f339800-8b9d-11ea-8898-c5f84a5a2333.png)
https://github.com/Web3Modal/web3modal#fortmatic

skylarweaver replied:
Which providers do we want to use @ligi?

ligi replied:
at least @walletconnect should be added IMHO. For me fortmatic (or every other provider that needs a key) can be removed IMHO

ligi replied:
perhaps we should add BurnerConnect for burner wallets

ligi commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -1,7 +1,143 @@
+window.addEventListener('load', async () => {
+  init();
+  document.querySelector("#btn-connect").addEventListener("click", onConnect);
+  document.querySelector("#btn-disconnect").addEventListener("click", onDisconnect);
```
do we have a disconnect button? Do not see much use in our case for it

pauljickling replied:
Agreed, will cut this out.

pauljickling committed:
Removed cruft, styled button, aremoved hardcoded value, and added warning text

pauljickling commented:
Added another commit that incorporated most of the feedback here. 

- Removed all (or nearly all of the copy/paste cruft).
- Added default CSS classes so the button looks like other buttons. This means that if you change values in the admin panel the button appearance will change as well!
- Added warning text below the button
- Targets input element without hard-coded id values (good because these values can change due to changes in `models.py`)

## Outstanding Issues
- Some of the inserted CSS values need more work (not responsive, and position of the added elements will look bad depending on browser width)
- Removed Fortmatic provider option, but want to make sure there is agreement on what providers we might add before diving into that

## Screenshot
<img width="1252" alt="wallet_address_button" src="https://user-images.githubusercontent.com/18646852/80845632-a6667100-8bbe-11ea-83a6-d0b59e9fe797.png">

skylarweaver commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -31,4 +86,26 @@ $(function () {
+  // TODO Better, responsive CSS values
+  warningText.style.cssText = "font-weight: 600; line-height: 1.5; margin-left: 300px;"
+  warningText.classList.add("warningText");
+  const warningTextContent = document.createTextNode("Please use a non-custodial address. E.g. not Coinbase, Gemini, etc.");
```
```suggestion
  const warningTextContent = document.createTextNode("Please use a non-custodial wallet address. E.g. not Coinbase, Binance, etc.");
```

pauljickling committed:
Update pretix_eth/static/pretix_eth/question-type.js

Co-authored-by: Skylar E. Weaver <weaver.skylar@gmail.com>

skylarweaver commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -1,7 +1,62 @@
+async function refreshAccountData() {
+  const web3 = new Web3(provider);
+  const accounts = await web3.eth.getAccounts();
+  const inputElements = document.querySelectorAll('input'); 
```
So if we add more inputs/questions to this form, this will get screwed up? We're definitely planning on adding more...

Is there a way to add an ID to this input and select it based on that? Maybe Jakub or Ligi would know if Pretix allows that when adding a question?

skylarweaver commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -1,7 +1,62 @@
+  const web3 = new Web3(provider);
+  const accounts = await web3.eth.getAccounts();
+  const inputElements = document.querySelectorAll('input'); 
+  /* IMPORTANT: length -2 because of a hidden input field at the end.
+   * If the input field for this page changes this value will likely need
+   * to be adjusted. */
```
So if we add more inputs/questions to this form, this will get screwed up? We're definitely planning on adding more...

Is there a way to add an ID to this input and select it based on that? Maybe Jakub or Ligi would know if Pretix allows that when adding a question?

kvbik replied:
`get_payment_eth_question_ids()` should be used for this, I believe.

kvbik replied:
Or of course feel free to rewrite that function to your needs. But there are the hidden fields marked with class `payment_eth_info` that contains comma separated list of IDs of desired address input question fields.

pauljickling replied:
Ok, just pushed a change to incorporate the function you wrote Jakub. Should be a stable query now.

skylarweaver commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -31,4 +86,26 @@ $(function () {
+  // Parent Element
+  const formGroupElements = document.querySelectorAll("div.form-group");
+  const lastFormGroupElement = formGroupElements.item(formGroupElements.length - 1);
+  
```
Can we add placeholder text to the input as follows: "Paste address here, or use below button to add address via Web3Modal"

Maybe we can even do this in Pretix @ligi ?

pauljickling committed:
Fixed CSS by placing new elements in correct div

pauljickling committed:
Merge branch 'question-plugin' of https://github.com/pauljickling/pretix-eth-payment-plugin into question-plugin

pauljickling committed:
Improved input document query

kvbik commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -1,7 +1,58 @@
+async function refreshAccountData() {
+  const web3 = new Web3(provider);
+  const accounts = await web3.eth.getAccounts();
+  const targetedInputElement = document.querySelector(`#${get_payment_eth_question_ids('payment_eth_info')}`);
```
You will have to iterate through this, they can multiple of them. Either if someone had created the address question multiple times in the admin (probably not intended) or if someone is purchasing more than one ticket (possible).

pauljickling committed:
Address button now handles use case of multiple tickets

ligi commented on pretix_eth/static/pretix_eth/question-type.js:
```diff
@@ -31,4 +85,27 @@ $(function () {
+
+  // Grandparent Element
+  const formGroupElements = document.querySelectorAll("div.form-group");
+  const lastFormGroupElement = formGroupElements.item(formGroupElements.length - 1);
```
we need to add a button into every of these form groups - not only the last one

pauljickling committed:
Creates multiple address buttons for multiple tickets, and targets specific input fields

pauljickling committed:
Replaced regex with function

kvbik commented:
thx for all the work @pauljickling if there is more work needed on this functionality, let's start new PRs.
